pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py

```diff
@@ -0,0 +1,184 @@
+            sources={"foo.md": "# Foo {#foo}\n## Bar {#bar}"},
+            redirects={"foo": ["foo.html"], "bar": ["foo.html"]},
+        )
+        before.raise_errors()
+
+        intermediate = self.setup_boilerplate(
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}"
+            },
+            redirects={"foo": ["foo.html"], "bar": ["foo.html"]},
+        )
+        self.assertRaises(InvalidCurrentPathError, lambda: intermediate.raise_errors(InvalidCurrentPathError))
+
+        after = self.setup_boilerplate(
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}"
+            },
+            redirects={"foo": ["foo.html"], "bar": ["bar.html", "foo.html#bar"]},
+        )
+        after.raise_errors()
+
+    def test_non_leaf_identifier_moved_to_different_file(self):
+        """Test moving a non-leaf identifier to a different source file."""
+        before = self.setup_boilerplate(
+            sources={"foo.md": "# Foo {#foo}\n## Bar {#bar}\n### Baz {#baz}"},
+            redirects={"foo": ["foo.html"], "bar": ["foo.html"], "baz": ["foo.html"]},
+        )
+        before.raise_errors()
+
+        intermediate = self.setup_boilerplate(
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}\n## Baz {#baz}"
+            },
+            redirects={"foo": ["foo.html"], "bar": ["foo.html"], "baz": ["foo.html"]},
+        )
+        self.assertRaises(InvalidCurrentPathError, lambda: intermediate.raise_errors(InvalidCurrentPathError))
+
+        after = self.setup_boilerplate(
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}\n## Baz {#baz}"
+            },
+            redirects={"foo": ["foo.html"], "bar": ["bar.html", "foo.html#bar"], "baz": ["bar.html", "foo.html#baz"]},
+        )
+        after.raise_errors()
+
+
+class TestGetClientRedirects(unittest.TestCase):
+    def test_no_client_redirects(self):
+        """Test fetching client side redirects and ignore server-side ones."""
+        redirects = Redirects(
+            {"foo": ["index.html"], "bar": ["index.html", "foo.html"]},
+            "const redirects = REDIRECTS_PLACEHOLDER;",
+        )
+        self.assertEqual(redirects.get_client_redirects("index.html"), "const redirects = {};")
+
+    def test_basic_redirect_matching(self):
+        redirects = Redirects(
+            {
+                'foo': ['index.html', 'index.html#some-section', 'index.html#another-section'],
+                'bar': ['index.html'],
+            },
+            "const redirects = REDIRECTS_PLACEHOLDER;",
+        )
+
+        result = redirects.get_client_redirects("index.html")
+        expected_redirects = {'some-section': 'index.html#foo', 'another-section': 'index.html#foo'}
+        expected_script = f"const redirects = {json.dumps(expected_redirects)};"
```

Comment:
Either `sort_keys=True` or read back the JSON into a dict. Or even better: let `get_client_redirects()` return JSON, and serialize that to a string at the call site.

Reply:
My concern with having `get_client_redirects()` return a dict is that it moves the responsibility of handling the substitution to `_file_header()` and it isn't really its job to do that? Or am I wrong to think that

Reply:
Why not? Why should the redirects be a data structure *baked into a string* instead of the actual data structure, at any point except where you absolutely need a string? The string should be the final representation that we don't touch any more, IMO.

Reply:
Fair enough.